### PR TITLE
[docs] Mentioned usage of custom daterange #453

### LINKED ISSUE #453: [docs] Mention custom date range usage in `List of endpoints` section
We recently added a date range widget to the project, which updates existing API endpoints (General & Device charts), so we need to update the [List of endpoints](https://github.com/openwisp/openwisp-monitoring#list-of-endpoints) section of the docs.

### PR DESCRIPTION
**Pull request by Aryamanz29**
Closes #453

Checks:

- [x] I have manually tested the proposed changes
- [ ] I have written new test cases to avoid regressions (if necessary)
- [x] I have updated the documentation (e.g. README.rst)

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -2344,15 +2344,25 @@ multi-tenancy and allows filtering monitoring data by ``organization_slug``,
 
     GET /api/v1/monitoring/dashboard/?organization_slug=<org1-slug>,<org2-slug>&location_id=<location1-id>,<location2-id>&floorplan_id=<floorplan1-id>,<floorplan2-id>
 
-Another parameter that can be used is ``time``, which allows to specify the
-time frame, the allowed values are:
+Another parameter that can be used is ``time``, which allows to specify the time frame
+or if you want to get general monitoring charts data for a specific date range then you can also add
+parameters ``start`` and  ``end`` to the API endpoint as shown below.
```
Comment (nemesisdesign) [inline README.rst:2349]: Are `start` and `end` an alternative to `time`? Does `time` have to be specified when `start` and `end` are supplied?

**Review by nemesisdesign — COMMENTED** [review]
It's still not clear to me, please see my question below.

File: README.rst (hunk 1 above)
Comment (Aryamanz29) [reply README.rst:2349]: Thanks for pointing this out, @nemesisdesign. Currently it is necessary to pass a `time` parameter with `start` and `end`, but I think a time parameter doesn't make sense in the case of a custom date range (we should calculate days after validating custom dates in views). I have updated the code in [`6623e56` (#450).](https://github.com/openwisp/openwisp-monitoring/pull/450/commits/6623e5635d200eec17ad82eba8fe8dffd55f88ec)

Comment (nemesisdesign) [reply README.rst]: I think it makes more sense to cherry-pick that commit into this patch and rename this patch to be a clarification and correction of this behavior. What do you think?

Comment (Aryamanz29) [reply README.rst]: Sure, Updated :+1:

File: openwisp_monitoring/views.py [hunk 2]
```diff
@@ -64,6 +64,13 @@ def get(self, request, *args, **kwargs):
         # if custom dates are provided then validate custom dates
         if start_date and end_date:
             self._validate_custom_date(start_date, end_date, timezone)
+            # if valid custom dates then calculate custom days
+            time = '1d'
+            end = datetime.strptime(end_date, '%Y-%m-%d %H:%M:%S')
+            start = datetime.strptime(start_date, '%Y-%m-%d %H:%M:%S')
+            custom_days = (end - start).days
+            if custom_days:
+                time = f'{custom_days}d'
```
Comment (nemesisdesign) [inline openwisp_monitoring/views.py:73]: I don't understand: why is this change needed?

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I did some changes to the text, please double check, see also my question below.

File: openwisp_monitoring/views.py (hunk 2 above)
Comment (Aryamanz29) [reply openwisp_monitoring/views.py:73]: Because we removed `time` from the query parameter, we must calculate the _`days diff (end - start)`_ after validating the custom date strings in the views in order to update [`custom_group_map`](https://github.com/openwisp/openwisp-monitoring/blob/ec248d1cee0cff8d81714d7c9d50adbab8f858f3/openwisp_monitoring/monitoring/base/models.py#L548-L577) accordingly.

Ref : https://github.com/openwisp/openwisp-monitoring/pull/454#discussion_r1020930821

Comment (nemesisdesign) [reply openwisp_monitoring/views.py:73]: ok thanks for the refersh.

I see that we are parsing the dates twice, which is bad.
We should modify the `_validate_custom_date` so that it returns the dates we need here, I would also rewrite this code as follows:

```python
start_datetime, end_datetime = self._validate_custom_date(start_date, end_date, timezone)
# if valid custom dates then calculate custom days
time = '1d'
custom_days = (start_datetime - end_datetime).days
if custom_days:
    time = f'{custom_days}d'
```

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: openwisp_monitoring/views.py (hunk 2 above)
Comment (Aryamanz29) [reply openwisp_monitoring/views.py:73]: thanks! @nemesisdesign

**Review by nemesisdesign — APPROVED** [review]
